pkg/mbedtls: initial pkg import to use entropy module

PeterKietzmann opened the pull request:
### Contribution description

This PR adds mbed TLS as a package to be used for entropy generation, and it is the only module that has been tested in RIOT so far. However, this PR is an enabler for additional mbed TLS features in RIOT.

The PR includes a simple RIOT API to access the entropy generated by mbed TLS. Entropy sources that feed the module are currently ADC Noise (#14324) and the HWRNG. Further information can be found in the [package documentation](https://github.com/PeterKietzmann/RIOT/blob/pr_add_pkg_mbedtls/pkg/mbedtls/doc.txt).


### Testing procedure
- Carefully read and understand the [high level documentation](https://github.com/PeterKietzmann/RIOT/blob/pr_add_pkg_mbedtls/pkg/mbedtls/doc.txt) (and of cause all other code & doc :) )
- Build doxygen (`make doc`).
- Execute the [provided test](https://github.com/PeterKietzmann/RIOT/tree/pr_add_pkg_mbedtls/tests/pkg_mbedtls) on boards and native.


### Issues/PRs references

- Successor of #14324
- Affected by #15660

miri64 commented on pkg/mbedtls/include/riot_config.h:
```diff
@@ -0,0 +1,3344 @@
+/**
+ * \file config.h
```
I might be mistaking, but this files seems similar to what I did with [`lwipopts.h` in lwIP](https://github.com/RIOT-OS/RIOT/blob/master/pkg/lwip/include/lwipopts.h). Does this file _has_ to be exactly this format or can we adapt it to be more RIOT-y? The fact, that this file is 96% comments, makes me think the latter.

I also don't think that you have to be that verbose on documentation, as the documentation (hopefully) lies with `mbedtls` (which you could just link here).

PeterKietzmann replied:
The file is a copy-paste from mbed TLS. For sure I could write a similar RIOT(-y) file, but for further integration of the package, I think it simplifies the development process to have a (almost) copy available. With that, a developer (i) sees immediately all available configuration options (&doc) and (ii) can easily look at diffs to the original file during version bump or the like...

miri64 replied:
Mh from the other point of view, I might be interested as a user, which configuration options of `mbedtls` are actually used. Scrolling though this 3300 line file to filter out the 142 lines that are actually important seems cumbersome, especially given, that you want to exclude this file from out `doxygen` build. Furthermore, we might want to make options configurable from the app (like we do e.g. with lwIP...),  which seems a bit more complicated when you need to comment out / comment in defines.

I also don't see the maintenance advantage you propose. If you want to see which options are available in `mbedtls`: go to the `mbedtls` documentation. If there is a version bump and config options change this will usually lead to failing compilation, giving you exactly the names of the macros that are new (and thus missing).

miri64 replied:
From https://github.com/RIOT-OS/RIOT/pull/15671#issuecomment-747636247

> Lets give it some more time and if no one opposes, we'll go with your proposal.

Yes, please @RIOT-OS/maintainers, can we have more thoughts and comments on this?

aabadie replied:
(I haven't checked the code, just followed the discussion)
If we can avoid the 3300 lines of configuration code, that would be better. If it's added, there's no reason to skip it from doxygen, or any other checks.
IIUC, the goal of this file is to provide the configuration of the package. Why not link it to the module mechanism of RIOT to define flags (like `MBEDTLS_SHA256_ALT`) or via a Kconfig file or both (depending on the cases) ?

- for the module version:
```c
ifdef MODULE_MBEDTLS_CIPHER_MODE_CBC
#define MBEDTLS_CIPHER_MODE_CBC
#endif
```

- for the Kconfig version (assuming there's a corresponding Kconfig boolean option):
```c
#if IS_ACTIVE(CONFIG_MBEDTLS_ECP_DP_SECP192R1_ENABLED)
#define MBEDTLS_ECP_DP_SECP192R1_ENABLED
#endif
```

PeterKietzmann commented:
To answer @miri64's question in [this comment](https://github.com/RIOT-OS/RIOT/issues/15660#issuecomment-747624626):

> And then there is the question why you want to exclude it in the first place? pkg/mbedtls/include/riot_config.h seems very RIOT-specific to me, so I guess it is not an external import. If it is: since you are in a pkg, can't you import it via the pkg mechanism as well?

It is an mbed TLS config file which I slightly adjusted for RIOT, by en-/disabling some options. By enabling additional library features, further changes might be required. To be still mostly "copy-paste-able" I've tried to keep changes at minimum.

Technically I could import & patch it. Practically, as a user, I would like the have the file directly available instead of a patch.

miri64 commented on pkg/mbedtls/contrib/entropy_mbedtls_riot.c:
```diff
@@ -0,0 +1,73 @@
+/*
```
Shouldn't this be its on module so in the future people can use `mbedtls` without the entropy source (if they so desire)?

So `pkg/mbedtls/contrib/mbedtls_entropy_source/...`. This way you also might not need to define pseudo-modules.

Again: see [lwIP](https://github.com/RIOT-OS/RIOT/blob/b45363b7844b6375239980d62d76811ae30b52a9/pkg/lwip/Makefile.include#L29-L37) how to implement actual modules within the `contrib` directory of a package.

PeterKietzmann replied:
Good point. I have tried this in my last commit, mind to have another look?

miri64 commented:
> It is an mbed TLS config file which I slightly adjusted for RIOT, by en-/disabling some options. By enabling additional library features, further changes might be required. To be still mostly "copy-paste-able" I've tried to keep changes at minimum.

Why do you want to have it "copy-paste-able"? Copy-paste from where? Is configuration so complicated within `mbedtls` that you need to copy files around to keep track of your configuration? I highly doubt any of this, given the popularity and porting status of the library.
 
> Technically I could import & patch it. Practically, as a user, I would like the have the file directly available instead of a patch.

PeterKietzmann commented:
>  Is configuration so complicated within mbedtls

Well, I guess it strongly depends on the point of view. A look at the amount of configuration options in that file gives you a first impression. Personally (proposing the library import here) I don't have a problem providing a specific configuration file. Still, I could think of other developers seeing this differently. Lets give it some more time and if no one opposes, we'll go with your proposal.

aabadie commented on pkg/mbedtls/Makefile:
```diff
@@ -0,0 +1,10 @@
+PKG_NAME=mbedtls
+PKG_URL=https://github.com/ARMmbed/mbedtls.git
+PKG_VERSION=2a1d9332d55d1270084232e42df08fdb08129f1b
```
If this commit hash correspond to a tagged version, please add it in a comment at the end of the line

aabadie commented on pkg/mbedtls/Makefile:
```diff
@@ -0,0 +1,10 @@
+
+all:
+	"$(MAKE)" -C $(PKGDIRBASE)/mbedtls/library/ \
+			  -f $(RIOTPKG)/mbedtls/Makefile.$(PKG_NAME)
```
`Makefile.mbedtls` only contains a minimal setup, you could just use Makefile.base directly and pass the MODULE variable in the command line:
```suggestion
			  -f $(RIOTBASE)/Makefile.base MODULE=$(PKG_NAME)
```
And Makefile.mbedtls can be removed.

aabadie commented on pkg/mbedtls/Makefile.dep:
```diff
@@ -0,0 +1,24 @@
+  # include sources if hardware is available
+  FEATURES_OPTIONAL += periph_hwrng
+  ifneq (,$(filter periph_hwrng,$(FEATURES_USED)))
+	  DEFAULT_MODULE += mbedtls_entropy_source_hwrng
```
indentation is off, it should be 2 spaces inside the ifneq block. Maybe it's a tab ?
Same comment applies in the rest of this file.

aabadie commented on pkg/mbedtls/Makefile.dep:
```diff
@@ -0,0 +1,24 @@
+  USEMODULE += mbedtls_contrib
+  USEMODULE += hashes
+
+  # include sources if hardware is available
+  FEATURES_OPTIONAL += periph_hwrng
+  ifneq (,$(filter periph_hwrng,$(FEATURES_USED)))
+	  DEFAULT_MODULE += mbedtls_entropy_source_hwrng
+  endif
+
+  FEATURES_OPTIONAL += periph_adc
+  ifneq (,$(filter periph_adc,$(FEATURES_USED)))
+	  DEFAULT_MODULE += mbedtls_entropy_source_adc
+  endif
```
What happens if the hardware provides both ? What is the entropy source selected in this case ?

leandrolanzieri replied:
I think those are not mutually exclusive, @PeterKietzmann ?

PeterKietzmann replied:
Then both are used (if not manually disabled). That is actually the intent of the entropy module, to accumulate multiple entropy sources.

aabadie commented on pkg/mbedtls/Makefile.dep:
```diff
@@ -0,0 +1,24 @@
+	  USEMODULE += entropy_source_adc_noise
+	  ifndef CONFIG_KCONFIG_USEMODULE_ENTROPY_SOURCE_ADC_NOISE
+		  # disable conditioning by default, enable health tests
+		  CFLAGS += -DCONFIG_ENTROPY_SOURCE_ADC_COND=0
+		  CFLAGS += -DCONFIG_ENTROPY_SOURCE_ADC_HEALTH_TEST=1
```
Those 2 constants seem to be unused.

PeterKietzmann replied:
Nope, they configure the ADC noise source which was added in #14324.

aabadie commented on pkg/mbedtls/Makefile.include:
```diff
@@ -0,0 +1,18 @@
+PSEUDOMODULES += mbedtls_entropy_source_adc
+
+ifneq (,$(filter mbedtls_contrib,$(USEMODULE)))
+    DIRS += $(RIOTPKG)/mbedtls/contrib
```
```suggestion
  DIRS += $(RIOTPKG)/mbedtls/contrib
```

aabadie commented on pkg/mbedtls/Makefile.include:
```diff
@@ -0,0 +1,18 @@
+endif
+
+ifneq (,$(filter mbedtls_entropy,$(USEMODULE)))
+    DIRS += $(RIOTPKG)/mbedtls/contrib/entropy
```
```suggestion
  DIRS += $(RIOTPKG)/mbedtls/contrib/entropy
```

aabadie commented on pkg/mbedtls/Makefile.mbedtls:
```diff
@@ -0,0 +1,3 @@
+MODULE = mbedtls
```
This file can be removed, see my previous comment.

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,122 @@
+#include "entropy_sources_mbedtls_riot.h"
+
+#include "kernel_defines.h"
+#if defined(MODULE_MBEDTLS_ENTROPY_SOURCE_HWRNG)
```
```suggestion
#if IS_USED(MODULE_MBEDTLS_ENTROPY_SOURCE_HWRNG)
```

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,122 @@
+#include "periph/hwrng.h"
+#endif
+
+#if defined(MODULE_MBEDTLS_ENTROPY_SOURCE_ADC)
```
```suggestion
#if IS_USED(MODULE_MBEDTLS_ENTROPY_SOURCE_ADC)
```

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,122 @@
+ * @{
+ */
+const entropy_source_mbedtls_riot_t entropy_funcs[] = {
+#ifdef MODULE_MBEDTLS_ENTROPY_SOURCE_HWRNG
```
You should use IS_USED here and in over places in this file

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,122 @@
+}
+
+#if defined(MODULE_MBEDTLS_ENTROPY_SOURCE_HWRNG)
+int riot_hwrng_poll( void *data, unsigned char *output, size_t len,
+                        size_t *olen )
```
```suggestion
int riot_hwrng_poll(void *data, unsigned char *output, size_t len,
                        size_t *olen)
```

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,122 @@
+#endif
+
+#if defined(MODULE_MBEDTLS_ENTROPY_SOURCE_ADC)
+int riot_adc_poll( void *data, unsigned char *output, size_t len,
+                    size_t *olen )
```
```suggestion
int riot_adc_poll(void *data, unsigned char *output, size_t len,
                    size_t *olen)
```

aabadie commented on pkg/mbedtls/contrib/sha256_alt.c:
```diff
@@ -0,0 +1,101 @@
+
+#include "sha256_alt.h"
+
+void mbedtls_sha256_init( mbedtls_sha256_context *ctx )
```
```suggestion
void mbedtls_sha256_init(mbedtls_sha256_context *ctx)
```

Same change required for all functions in this file

aabadie commented on pkg/mbedtls/contrib/sha256_alt.c:
```diff
@@ -0,0 +1,101 @@
+
+#include "riot_config.h"
+
+#if defined(MBEDTLS_SHA256_ALT)
```
```suggestion
#if IS_ACTIVE(MBEDTLS_SHA256_ALT)
```
(and add include of kernel_defines.h)

aabadie commented on pkg/mbedtls/doc.txt:
````diff
@@ -0,0 +1,101 @@
+ *
+ * ```
+ * FEATURES_PROVIDED += periph_hwrng
+ * FEATURES_PROVIDED += entropy_source_adc_noise
````
```suggestion
 * FEATURES_PROVIDED += periph_adc
```
?

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,122 @@
+ */
+
+/**
+ * @ingroup     pkg_mbedtls
```
```suggestion
 * @ingroup     pkg_mbedtls_entropy
```
?

PeterKietzmann replied:
Yes, probably a good idea to spend an own doxygen group now that we moved the entropy support to its own module.

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_mbedtls_riot.c:
```diff
@@ -0,0 +1,73 @@
+ */
+
+/**
+ * @ingroup     pkg_mbedtls
```
```suggestion
 * @ingroup     pkg_mbedtls_entropy
```
?

aabadie commented on pkg/mbedtls/include/sha256_alt.h:
```diff
@@ -0,0 +1,45 @@
+extern "C" {
+#endif
+
+#if defined (MBEDTLS_SHA256_ALT)
```
```suggestion
#if IS_ACTIVE(MBEDTLS_SHA256_ALT)
```

aabadie commented on tests/pkg_mbedtls/Makefile:
```diff
@@ -0,0 +1,15 @@
+
+include $(RIOTBASE)/Makefile.include
+
+ifneq (,$(filter mbedtls_entropy_source_adc,$(USEMODULE)))
+	ifndef CONFIG_KCONFIG_USEMODULE_ENTROPY_SOURCE_ADC_NOISE
+		CFLAGS += -DCONFIG_ENTROPY_SOURCE_ADC_HMIN=65536
+	endif
+endif
```
Why don't you put this in the package Makefile.include ?

PeterKietzmann replied:
To avoid mis-configuration of this entropy source, we set an invalid default value. Hence, a user must evaluate and configure the module according to the hardware in use and the deployment. This line prevents failures during testing.

aabadie reviewed (changes requested):
Went through this PR and have some comment related to the build system and other minor things.
I also think that the riot_config.h should be rewritten and not excluded from the doxygen build (since it might contain important configuration information) and the CI checks.

PeterKietzmann commented:
@aabadie many thanks for your review. I have left some answers inline and will work on the other comments. Regarding the configuration file there seems to be a favor for the RIOT-style option. Lets give it some time anyway, and I will provide an alternative during the next couple of days if no one objects.

PeterKietzmann commented:
@aabadie I have addressed your comments (despite the config header file as described above)

fjmolinas commented:
ping @aabadie

leandrolanzieri commented on pkg/mbedtls/Kconfig:
```diff
@@ -0,0 +1,46 @@
+config MBEDTLS_ENTROPY_C
+    bool "Enable the platform-specific entropy code."
+    default y
+    depends on MBEDTLS_SHA256_C || MBEDTLS_SHA512_C
```
Where is `MBEDTLS_SHA512_C` being defined?

PeterKietzmann replied:
We do not expose it right now, a future extension could do so. I kept it here as a reminder. Furthermore, the documentation of mbedtls clearly states that "entropy" depends on sha256 or sha512

leandrolanzieri replied:
If you really think we need to keep it there for a reference, I'd do: `depends on MBEDTLS_SHA256_C # || MBEDTLS_SHA512_C`

PeterKietzmann replied:
done

leandrolanzieri commented on pkg/mbedtls/contrib/entropy/Kconfig:
```diff
@@ -0,0 +1,41 @@
+        first argument.
+
+config MBEDTLS_NO_PLATFORM_ENTROPY
+    bool "Enable this macro to disable the built-in platform entropy functions"
```
```suggestion
    bool "Disable the built-in platform entropy functions"
```

leandrolanzieri commented on pkg/mbedtls/include/riot_config.h:
```diff
@@ -57,20 +59,24 @@
  */
+#if IS_ACTIVE(CONFIG_MBEDTLS_SHA256_ALT)
 #define MBEDTLS_SHA256_ALT 1
+#endif
```
I think this may cause confusion. If the exposed configuration parameters to the user are the `CONFIG_` ones, then that's what should be documented in Doxygen. So in this case the user should find `CONFIG_MBEDTLS_SHA256_ALT` instead of `MBEDTLS_SHA256_ALT`.

Additionally, as all the configurations seem to default to `1`, we may need to guard them around a check for `CONFIG_KCONFIG_USEMODULE_MBEDTLS`, because `false` in Kconfig translates to a not defined macro.

PeterKietzmann commented:
@leandrolanzieri thanks for taking a look. I've pushed some commits

PeterKietzmann commented:
Added Kconfig options, enabled config header for doxygen again, tested on native and nrf52840dk, cleaned up minor stuff...

@aabadie, mind to give it an other look to remove the change request?

PeterKietzmann commented:
Btw: all tests now pass

fjmolinas commented:
@leandrolanzieri is this one ready to go?

leandrolanzieri commented on pkg/mbedtls/include/entropy_mbedtls_riot.h:
```diff
@@ -0,0 +1,82 @@
+ * A context structure is allocated and managed internally.
+ * Connect all available entropy sources to the poll.
+ *
+ * @return    0 if successful
+ * @return    MBEDTLS_ERR_ENTROPY_MAX_SOURCES on failure
```
Here and bellow:
```suggestion
 * @retval    0 if successful
 * @retval    MBEDTLS_ERR_ENTROPY_MAX_SOURCES on failure
```

leandrolanzieri commented on pkg/mbedtls/include/entropy_mbedtls_riot.h:
```diff
@@ -0,0 +1,82 @@
+ *
+ * @pre Module is initialized already.
+ *
+ * @param[in] output   Pointer to the output buffer to fill with entropy values.
```
```suggestion
 * @param[out] output   Pointer to the output buffer to fill with entropy values.
```

leandrolanzieri commented on pkg/mbedtls/include/entropy_mbedtls_riot.h:
```diff
@@ -0,0 +1,82 @@
+/**
+ * @brief   Retrieve entropy values after initialization.
+ *
+ * @pre Module is initialized already.
```
```suggestion
 * @pre Module is initialized already (i.e. @ref entropy_mbedtls_riot_init has been called).
```

leandrolanzieri commented on pkg/mbedtls/include/entropy_mbedtls_riot.h:
```diff
@@ -0,0 +1,82 @@
+ * internally. Requested entropy values are gathered and the context is
+ * uninitializedafterwards.
+ *
+ * @param[in] output   Pointer to the output buffer to fill with entropy values.
```
```suggestion
 * @param[out] output   Pointer to the output buffer to fill with entropy values.
```

leandrolanzieri commented on pkg/mbedtls/include/entropy_mbedtls_riot.h:
```diff
@@ -0,0 +1,82 @@
+ *
+ * Convenience function. A context structure is allocated and managed
+ * internally. Requested entropy values are gathered and the context is
+ * uninitializedafterwards.
```
```suggestion
 * uninitialized afterwards.
```

leandrolanzieri commented on pkg/mbedtls/include/entropy_sources_mbedtls_riot.h:
```diff
@@ -0,0 +1,93 @@
+ */
+typedef struct {
+    mbedtls_entropy_f_source_ptr func; /**< Pointer to entropy callback func. */
+    int strong;                        /**< Strength of the entropy function
+                                         *  (strong/weak) */
```
Is this 1 or 0? Could we use an enum?

PeterKietzmann replied:
Yes, only 0/1 (`MBEDTLS_ENTROPY_SOURCE_WEAK`/`MBEDTLS_ENTROPY_SOURCE_STRONG`). In practice, however, this resembles mbed TLS logic. I would like to avoid translating this logic for the RIOT adaptation.

leandrolanzieri replied:
This needs *at least* documentation on the possible values then.

leandrolanzieri commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,123 @@
+ */
+int riot_add_entropy_src_avail(mbedtls_entropy_context *ctx)
+{
+    int ret = 0;
```
```suggestion
    assert(ctx);
    int ret = 0;
```

leandrolanzieri commented on tests/pkg_mbedtls/Makefile:
```diff
@@ -0,0 +1,55 @@
+include ../Makefile.tests_common
+
+# The following boards currently miss entropy sources.
+BOARD_BLACKLIST :=  6lowpan-clicker \
+                    atxmega-a1u-xpro \
+                    atxmega-a3bu-xplained \
+                    avsextrem \
+                    cc1312-launchpad \
+                    cc1350-launchpad \
+                    cc1352-launchpad \
+                    cc1352p-launchpad \
+                    cc2650-launchpad \
+                    cc2650stk \
+                    fox \
+                    hifive1 \
+                    hifive1b \
+                    iotlab-a8-m3 \
+                    iotlab-m3 \
+                    limifrog-v1 \
+                    lobaro-lorabox \
+                    maple-mini \
+                    mbed_lpc1768 \
+                    msb-430 \
+                    msb-430h \
+                    nucleo-f103rb \
+                    nucleo-f446ze \
+                    nucleo-g070rb \
+                    nucleo-g071rb \
+                    nucleo-wl55jc \
+                    opencm904 \
+                    seeeduino_arch-pro \
+                    spark-core \
+                    stm32f3discovery \
+                    stm32mp157c-dk2 \
+                    teensy31 \
+                    telosb \
+                    z1 \
+                    #
```
Why is this needed? Wouldn't the boards with missing `periph_adc` and `periph_hwrng` just be ignored by not meeting the `FEATURES_REQUIRED`?

PeterKietzmann replied:
These are no hard feature requirements. This is why we made them `FEATURES_OPTIONAL` in the Makefile.dep. The boards listed have neither and ADC (exposed) nor a HWRNG

leandrolanzieri replied:
Could you check if the following works?
```
FEATURES_REQUIRED_ANY += periph_adc|periph_hwrng
FEATURES_OPTIONAL += periph_adc periph_hwrng
```

The first line checks for at least one of them, but prioritizes the first one (i.e. when both are present `periph_hwrng` would not be used). The second line tries to enable both if provided. When none is provided this would fail.

leandrolanzieri commented on tests/pkg_mbedtls/Makefile.ci:
```diff
@@ -0,0 +1,16 @@
+    nucleo-l011k4 \
+    nucleo-l476rg \
+    qn9080dk \
+    #
```
Missing newline

PeterKietzmann commented on pkg/mbedtls/contrib/entropy/entropy_mbedtls_riot.c:
```diff
@@ -0,0 +1,73 @@
+#include "entropy_sources_mbedtls_riot.h"
+#include "entropy_mbedtls_riot.h"
+
+/* Internal mbedtls entropy context for convenience functions */
+static mbedtls_entropy_context ctx;
```
`mbedtls_entropy_context` holds a mutex if mbedTLS is configured in threading mode (see [here](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/entropy.h#L133). Maybe we should rather expose that mode with Kconfig.

leandrolanzieri replied:
OK, I'd enable it by default. I think you should also add it to the configuration header file.

PeterKietzmann replied:
I've added the threading mode to this pkg port.

aabadie commented on pkg/mbedtls/Makefile:
```diff
@@ -0,0 +1,10 @@
+include $(RIOTBASE)/pkg/pkg.mk
+
+all:
+	"$(MAKE)" -C $(PKGDIRBASE)/mbedtls/library/ \
```
```suggestion
	"$(QQ)$(MAKE)" -C $(PKGDIRBASE)/mbedtls/library/ \
```

PeterKietzmann commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,122 @@
+int mbedtls_hardware_poll(void *data, unsigned char *output, size_t len,
+                            size_t *olen)
+{
+    assert(output != NULL && olen!= NULL);
```
yes

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_mbedtls_riot.c:
```diff
@@ -0,0 +1,73 @@
+
+int entropy_mbedtls_riot_init(void)
+{
+    int ret = 0;
+
+    mbedtls_entropy_init(&ctx);
+
+    ret = riot_add_entropy_src_avail(&ctx);
+
+    return ret;
```
```suggestion
    mbedtls_entropy_init(&ctx);

    return riot_add_entropy_src_avail(&ctx);;
```

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_mbedtls_riot.c:
```diff
@@ -0,0 +1,73 @@
+{
+    assert(output != NULL);
+
+    int ret;
+
+    ret = mbedtls_entropy_func(&ctx, output, len);
+
+    return ret;
```
```suggestion
    return mbedtls_entropy_func(&ctx, output, len);;
```

aabadie commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,123 @@
+    entropy_source_adc_init();
+#endif
+
+    for (unsigned i = 1; i < NUM_ENTROPY_FUNCS; i++) {
+        ret = mbedtls_entropy_add_source(ctx, entropy_funcs[i].func, NULL,
+                                1, entropy_funcs[i].strong);
+    }
```
I think there's something wrong with the ret usage. For example, if something goes wrong with the first entropy source (ret != 0) but works with the last, the whole function return code will be 0. I don't think this is expected.

Maybe it's better to check the return value in the for loop and return with an error if any of source cannot be added ?

PeterKietzmann replied:
The loop adds entropy sources to a list which will be polled during entropy gathering. The error case occurs when a maximum number of sources have been added. Hence, the case of failure followed by success in not true. Nevertheless, we can save some redundant calls by returning on the first error.

leandrolanzieri commented on pkg/mbedtls/Kconfig:
```diff
@@ -0,0 +1,46 @@
+# This file is subject to the terms and conditions of the GNU Lesser
+# General Public License v2.1. See the file LICENSE in the top level
+# directory for more details.
+
```
Hmm actually the entropy sources are not modelled in master yet. Ok, let's leave for a follow up then...

PeterKietzmann commented:
@aabadie , @leandrolanzieri your latest comments should be addressed in this commit 843c4971345e0f77e7d6d02acb140d21f7108af1

PeterKietzmann commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,122 @@
+
+#if defined(MODULE_MBEDTLS_ENTROPY_SOURCE_ADC)
+    entropy_source_adc_init();
+#endif
```
It depends on the source, but seems likely that most sources have some kind of an init function. Technically, we could add  .init function to `entropy_source_mbedtls_riot_t` but I would like to delay that, since other sources (e.g. PUF) might have very different requirements from ADC sampling or HWRNG readout.

benpicco commented on pkg/mbedtls/Makefile:
```diff
@@ -0,0 +1,10 @@
+PKG_NAME=mbedtls
+PKG_URL=https://github.com/ARMmbed/mbedtls.git
+PKG_VERSION=v2.26.0
```
Why not use v3.0.0 here? (or at least v2.27.0)

benpicco commented on pkg/mbedtls/contrib/entropy/entropy_sources_mbedtls_riot.c:
```diff
@@ -0,0 +1,127 @@
+
+    for (unsigned i = 1; i < NUM_ENTROPY_FUNCS; i++) {
+        ret = mbedtls_entropy_add_source(ctx, entropy_funcs[i].func, NULL,
+                                1, entropy_funcs[i].strong);
```
```suggestion
                                         1, entropy_funcs[i].strong);
```

benpicco commented on tests/pkg_mbedtls/Makefile:
```diff
@@ -0,0 +1,18 @@
+
+ifneq (,$(filter mbedtls_entropy_source_adc,$(USEMODULE)))
+  ifndef CONFIG_KCONFIG_USEMODULE_ENTROPY_SOURCE_ADC_NOISE
+    CFLAGS += -DCONFIG_ENTROPY_SOURCE_ADC_HMIN=65536
```
Would be good to have a comment where that value comes from

benpicco commented on pkg/mbedtls/include/entropy_sources_mbedtls_riot.h:
```diff
@@ -0,0 +1,93 @@
+ *                     MBEDTLS_ERR_ENTROPY_SOURCE_FAILED otherwise.
+ */
+int riot_adc_poll(void *data, unsigned char *output, size_t len,
+                    size_t *olen);
```
```suggestion
                  size_t *olen);
```

benpicco commented on pkg/mbedtls/include/entropy_sources_mbedtls_riot.h:
```diff
@@ -0,0 +1,93 @@
+ * @return             0 always
+ */
+int riot_hwrng_poll(void *data, unsigned char *output, size_t len,
+                        size_t *olen);
```
```suggestion
                     size_t *olen);
```

benpicco commented on pkg/mbedtls/include/entropy_sources_mbedtls_riot.h:
```diff
@@ -0,0 +1,93 @@
+    mbedtls_entropy_f_source_ptr func; /**< Pointer to entropy callback func. */
+    int strong;                        /**< Strength of the entropy function
+                                         *  (strong=1 for high entropy sources,
+                                         *   strong=0 for weak entropy sources)*/
```
If it only has two states, why not `bool`?

PeterKietzmann replied:
1. This is what [mbedtls](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/entropy.h#L182) expects.
2. I personally think that they left themselves a backdoor to pass an entropy value instead of only strong/weak.

benpicco reviewed (approved):
Please squash!

fjmolinas commented:
@PeterKietzmann the test fails on `samr-21xpro` (and a couple of other BOARDs I tested locally, some report a "failed", others hang), is that expected? The PR only mentioned testing on `native`, is it supposed to work on hardware?

PeterKietzmann commented:
> is it supposed to work on hardware

Yes it is. I ran the test on different platforms, back then. Cannot remember any high-ups. Just wondering: who has tested this PR before merging? Anyway, I will look into this as soon as I can!

PeterKietzmann commented:
> some report a "failed", others hang),

@fjmolinas can you name these boards so I have a starting point?

fjmolinas commented:
> > some report a "failed", others hang),
> 
> @fjmolinas can you name these boards so I have a starting point?

[Nightlies](https://ci.riot-os.org/nightlies.html) are failing on `samr21-xpro`, I got a hang on `nucleo-l152re`. I'll run the test on all the BOARDs I have to give a complete list, will do on this PR and master.

fjmolinas commented:
I ran the tests on the [PR](https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/307/artifact/) and on [master](https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/308/artifact/).

Results (ignore failed compilations, there is some kind of raise condition with pkg downloads...):

<details><summary>master</summary>

```
### nucleo-g071rb/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](nucleo-g071rb/tests/pkg_mbedtls/compilation.failed)

### nucleo-f207zg/failuresummary


### nucleo-f334r8/failuresummary


### nucleo-g474re/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](nucleo-g474re/tests/pkg_mbedtls/compilation.failed)

### nrf52dk/failuresummary


### frdm-k64f/failuresummary


### nucleo-l152re/failuresummary

Failures during test:
- [tests/pkg_mbedtls](nucleo-l152re/tests/pkg_mbedtls/test.failed)

### arduino-uno/failuresummary


### nucleo-f767zi/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](nucleo-f767zi/tests/pkg_mbedtls/compilation.failed)

### samr21-xpro/failuresummary

Failures during test:
- [tests/pkg_mbedtls](samr21-xpro/tests/pkg_mbedtls/test.failed)

### p-nucleo-wb55/failuresummary


### nucleo-l496zg/failuresummary


### nucleo-l452re/failuresummary


### frdm-kw41z/failuresummary

Failures during test:
- [tests/pkg_mbedtls](frdm-kw41z/tests/pkg_mbedtls/test.failed)

### nucleo-l433rc/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](nucleo-l433rc/tests/pkg_mbedtls/compilation.failed)

### i-nucleo-lrwan1/failuresummary


### openmote-b/failuresummary


### nucleo-wl55jc/failuresummary


### nucleo-l4r5zi/failuresummary


### z1/failuresummary


### nucleo-f091rc/failuresummary

Failures during test:
- [tests/pkg_mbedtls](nucleo-f091rc/tests/pkg_mbedtls/test.failed)

### arduino-zero/failuresummary

Failures during test:
- [tests/pkg_mbedtls](arduino-zero/tests/pkg_mbedtls/test.failed)

### nucleo-l552ze-q/failuresummary


### nucleo-f303k8/failuresummary


### nucleo-f103rb/failuresummary


### b-l072z-lrwan1/failuresummary


### atmega256rfr2-xpro/failuresummary

Failures during test:
- [tests/pkg_mbedtls](atmega256rfr2-xpro/tests/pkg_mbedtls/test.failed)

### nucleo-f303re/failuresummary


### slstk3402a/failuresummary


### nucleo-f030r8/failuresummary

Failures during test:
- [tests/pkg_mbedtls](nucleo-f030r8/tests/pkg_mbedtls/test.failed)

### nucleo-l073rz/failuresummary

```
</details>

<details><summary>pr</summary>

```

### nucleo-g071rb/failuresummary

Failures during test:
- [tests/pkg_mbedtls](nucleo-g071rb/tests/pkg_mbedtls/test.failed)

### nucleo-f207zg/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](nucleo-f207zg/tests/pkg_mbedtls/compilation.failed)

### nucleo-f334r8/failuresummary


### nucleo-g474re/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](nucleo-g474re/tests/pkg_mbedtls/compilation.failed)

### nrf52dk/failuresummary


### frdm-k64f/failuresummary


### nucleo-l152re/failuresummary

Failures during test:
- [tests/pkg_mbedtls](nucleo-l152re/tests/pkg_mbedtls/test.failed)

### arduino-uno/failuresummary


### nucleo-f767zi/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](nucleo-f767zi/tests/pkg_mbedtls/compilation.failed)

### samr21-xpro/failuresummary

Failures during test:
- [tests/pkg_mbedtls](samr21-xpro/tests/pkg_mbedtls/test.failed)

### p-nucleo-wb55/failuresummary


### nucleo-l496zg/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](nucleo-l496zg/tests/pkg_mbedtls/compilation.failed)

### nucleo-l452re/failuresummary


### frdm-kw41z/failuresummary

Failures during test:
- [tests/pkg_mbedtls](frdm-kw41z/tests/pkg_mbedtls/test.failed)

### nucleo-l433rc/failuresummary


### i-nucleo-lrwan1/failuresummary


### openmote-b/failuresummary


### nucleo-wl55jc/failuresummary


### nucleo-l4r5zi/failuresummary


### z1/failuresummary


### nucleo-f091rc/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](nucleo-f091rc/tests/pkg_mbedtls/compilation.failed)

### arduino-zero/failuresummary

Failures during compilation:
- [tests/pkg_mbedtls](arduino-zero/tests/pkg_mbedtls/compilation.failed)

### nucleo-l552ze-q/failuresummary


### nucleo-f303k8/failuresummary


### nucleo-f103rb/failuresummary


### b-l072z-lrwan1/failuresummary


### atmega256rfr2-xpro/failuresummary

Failures during test:
- [tests/pkg_mbedtls](atmega256rfr2-xpro/tests/pkg_mbedtls/test.failed)

### nucleo-f303re/failuresummary


### slstk3402a/failuresummary


### nucleo-f030r8/failuresummary

Failures during test:
- [tests/pkg_mbedtls](nucleo-f030r8/tests/pkg_mbedtls/test.failed)

### nucleo-l073rz/failuresummary
```

</details>

PeterKietzmann commented:
Just a quick heads-up:
- The failure on `samr21-xpro` is actually a good thing. It has insufficient entropy sources and configures only a fallback ADC line to gather noise using the `adc_noise` module. This detects failures during the runtime tests which lets the (mbedtls) entropy module fail. I will improve the test to better reflect this behavior.
- Hanging on `nucleo-l152re` is due to a missing ADC initialization before use. Some ADC drivers block on sampling before initialization. Turns our, I forgot to call the initialization function in `auto_init`. Will add it.
- Lets see if these changes will already let the other boards pass.
